go/vt/vtgate/engine/insert_select.go

```diff
@@ -0,0 +1,419 @@
+		result.InsertID = uint64(insertID)
+	}
+	return result, nil
+}
+
+func (ins *InsertSelect) getInsertQueries(
+	ctx context.Context,
+	vcursor VCursor,
+	bindVars map[string]*querypb.BindVariable,
+	rows []sqltypes.Row,
+) ([]*srvtopo.ResolvedShard, []*querypb.BoundQuery, error) {
+	colVindexes := ins.ColVindexes
+	if len(colVindexes) != len(ins.VindexValueOffset) {
+		return nil, nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "vindex value offsets and vindex info do not match")
+	}
+
+	// Here we go over the incoming rows and extract values for the vindexes we need to update
+	vindexRowsValues := make([][]sqltypes.Row, len(colVindexes))
+	for _, inputRow := range rows {
+		for colIdx := range colVindexes {
+			offsets := ins.VindexValueOffset[colIdx]
+			row := make(sqltypes.Row, 0, len(offsets))
+			for _, offset := range offsets {
+				if offset == -1 { // value not provided from select query
+					row = append(row, sqltypes.NULL)
+					continue
+				}
+				row = append(row, inputRow[offset])
+			}
+			vindexRowsValues[colIdx] = append(vindexRowsValues[colIdx], row)
+		}
+	}
+
+	keyspaceIDs, err := ins.processVindexes(ctx, vcursor, vindexRowsValues, colVindexes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var indexes []*querypb.Value
+	var destinations []key.Destination
+	for i, ksid := range keyspaceIDs {
+		if ksid != nil {
+			indexes = append(indexes, &querypb.Value{
+				Value: strconv.AppendInt(nil, int64(i), 10),
+			})
+			destinations = append(destinations, key.DestinationKeyspaceID(ksid))
+		}
+	}
+	if len(destinations) == 0 {
+		// In this case, all we have is nil KeyspaceIds, we don't do
+		// anything at all.
+		return nil, nil, nil
+	}
+
+	rss, indexesPerRss, err := vcursor.ResolveDestinations(ctx, ins.Keyspace.Name, indexes, destinations)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	queries := make([]*querypb.BoundQuery, len(rss))
+	for i := range rss {
+		bvs := sqltypes.CopyBindVariables(bindVars) // we don't want to create one huge bindvars for all values
+		var mids sqlparser.Values
+		for _, indexValue := range indexesPerRss[i] {
+			index, _ := strconv.Atoi(string(indexValue.Value))
+			if keyspaceIDs[index] != nil {
+				row := sqlparser.ValTuple{}
+				for colOffset, value := range rows[index] {
+					bvName := insertVarOffset(index, colOffset)
+					bvs[bvName] = sqltypes.ValueBindVariable(value)
+					row = append(row, sqlparser.NewArgument(bvName))
+				}
+				mids = append(mids, row)
+			}
+		}
+		rewritten := ins.Prefix + sqlparser.String(mids) + ins.Suffix
+		queries[i] = &querypb.BoundQuery{
+			Sql:           rewritten,
+			BindVariables: bvs,
+		}
+	}
+
+	return rss, queries, nil
+}
+
+func (ins *InsertSelect) execInsertSharded(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	result, err := ins.execSelect(ctx, vcursor, bindVars)
+	if err != nil {
+		return nil, err
+	}
+	if len(result.rows) == 0 {
+		return &sqltypes.Result{}, nil
```

Comment:
shouldn't we pass on the incoming result instead? there might be other metadata that should be sent along, no?

Reply:
the result is of the `select statement`, the final result should be of `insert` hence if there are no rows we need to return an empty result set. like fields are irrelevant for insert statements, will check if anything other metadata can be used.